cachix-api/src/Cachix/Api/Signing.hs

```diff
@@ -18,11 +19,15 @@ import qualified Data.Text as T
 import           Data.Text (Text)
 
 
+data FingerprintInvalid =
+  NarSizeIsZero
+  deriving (Show)
+
 -- perl/lib/Nix/Manifest.pm:fingerprintPath
--- TODO: Either Text ByteString: assert values
 -- NB: references must be sorted
-fingerprint :: Text -> Text -> Int -> [Text] -> ByteString
-fingerprint storePath narHash narSize references = toS $ T.intercalate ";"
+fingerprint :: Text -> Text -> Int -> [Text] -> Either FingerprintInvalid ByteString
```

Comment:
Not sure why fingerprinting should fail. It's the narinfo that's invalid, not that the fingerprint can't be made. Isn't there someplace else where you could also validate the store path, hash format etc?